File: Problem5.java
// Time Complexity : O(m*n), where m is no of rows and n is no. of columns
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix.length == 0 || matrix[0].length == 0) return false;

        int col = matrix[0].length;
        int start = 0;
        // end would be 11 for 3*4 as start = 0
        int end = matrix.length * matrix[0].length - 1;


        while(start <= end){
            int mid = start + (end - start) / 2;
            int element = matrix[mid / col][mid % col];

            if(target > element){
                start = mid + 1;
            }else if(target < element){
                end = mid - 1;
            }else{
                return true;
            }
        }

        return false;
    }
}

File: Problem5.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        lo, hi = 0, len(matrix) - 1
        while lo <= hi:
            midRow = (lo + hi) // 2
            row = matrix[midRow]
            if target < row[0]: hi = midRow - 1
            elif target > row[-1]: lo = midRow + 1
            else:
                l, h = 0, len(row) - 1
                while l <= h:
                    mid = (l + h) // 2
                    if row[mid] == target: return True
                    elif target > row[mid]: l = mid + 1
                    else: h = mid - 1
                return False
 
 #Time Complexity : O(logM) + O(logN)
 #Used binary search for sorting elements 
 # to check elements in each row low to high